File: benchmarks/regex-redux.js
let t0 = Date.now();

const seq = "agggtaaa cgtgggtaaa aactggtaaa agactgtaaa aggacttaaa agggcgaaa agggtcaa agggtaca agggtaac aatggtaaa agagtaaa aggataaa agggcaaa tHaNt aND caN HaD WaS aNt BY <header> |word| ";

let sb = "";
for (let i = 0; i < 250; i++) {
    sb += ">Sequence_" + i + "\n" + seq + "\n";
}
let text = sb;
let originalLen = Buffer.byteLength(text, 'utf8');

text = text.replace(/(>[^\n]+)?\n/g, "");
let cleanedLen = Buffer.byteLength(text, 'utf8');

const variants = [
    /agggtaaa|tttaccct/g,
    /[cgt]gggtaaa|tttaccc[acg]/g,
    /a[act]ggtaaa|tttacc[agt]t/g,
    /ag[act]gtaaa|tttac[agt]ct/g,
    /agg[act]taaa|ttta[agt]cct/g,
    /aggg[acg]aaa|ttt[cgt]ccct/g,
    /agggt[cgt]aa|tt[acg]accct/g,
    /agggta[cgt]a|t[acg]taccct/g,
    /agggtaa[cgt]|[acg]ttaccct/g
];

for (let i = 0; i < variants.length; i++) {
    const re = variants[i];
    const match = text.match(re);
    const count = match ? match.length : 0;
    // get string rep of regex without flags
    let reStr = re.toString();
    reStr = reStr.substring(1, reStr.length - 2); 
    console.log(reStr + " " + count);
}

const substitutions = [
    [/tHa[Nt]/g, "<4>"],
    [/aND|caN|Ha[DS]|WaS/g, "<3>"],
    [/a[NSt]|BY/g, "<2>"],
    [/<[^>]*>/g, "|"],
    [/\|[^|][^|]*\|/g, "-"]
];

for (let i = 0; i < substitutions.length; i++) {
    text = text.replace(substitutions[i][0], substitutions[i][1]);
}

console.log("\n" + originalLen + "\n" + cleanedLen + "\n" + Buffer.byteLength(text, 'utf8'));
console.log("bench_ns=" + ((Date.now() - t0) * 1000000));

File: benchmarks/regex-redux.rs
use regex::Regex;
use std::fmt::Write;
use std::time::Instant;

fn main() {
    let t0 = Instant::now();

    let seq = "agggtaaa cgtgggtaaa aactggtaaa agactgtaaa aggacttaaa agggcgaaa agggtcaa agggtaca agggtaac aatggtaaa agagtaaa aggataaa agggcaaa tHaNt aND caN HaD WaS aNt BY <header> |word| ";

    let mut sb = String::new();
    for i in 0..250 {
        write!(&mut sb, ">Sequence_{}\n{}\n", i, seq).unwrap();
    }
    let mut text = sb;
    let original_len = text.len();

    let clean_re = Regex::new(r"(>[^\n]+)?\n").unwrap();
    text = clean_re.replace_all(&text, "").into_owned();
    let cleaned_len = text.len();

    let variants = vec![
        "agggtaaa|tttaccct",
        "[cgt]gggtaaa|tttaccc[acg]",
        "a[act]ggtaaa|tttacc[agt]t",
        "ag[act]gtaaa|tttac[agt]ct",
        "agg[act]taaa|ttta[agt]cct",
        "aggg[acg]aaa|ttt[cgt]ccct",
        "agggt[cgt]aa|tt[acg]accct",
        "agggta[cgt]a|t[acg]taccct",
        "agggtaa[cgt]|[acg]ttaccct",
    ];

    for v in variants {
        let re = Regex::new(v).unwrap();
        let count = re.find_iter(&text).count();
        println!("{} {}", v, count);
    }

    let substitutions = vec![
        ("tHa[Nt]", "<4>"),
        ("aND|caN|Ha[DS]|WaS", "<3>"),
        ("a[NSt]|BY", "<2>"),
        ("<[^>]*>", "|"),
        (r"\|[^|][^|]*\|", "-"),
    ];

    for (re_str, sub) in substitutions {
        let re = Regex::new(re_str).unwrap();
        text = re.replace_all(&text, sub).into_owned();
    }

    println!("\n{}\n{}\n{}", original_len, cleaned_len, text.len());
    let duration = t0.elapsed();
    println!("bench_ns={}", duration.as_nanos());
}
